### PR TITLE
AC_PosControl: move PID objects into PosControl library

### PR DESCRIPTION
**Pull request by rmackay9**
This PR includes the following changes which move the acceleration, velocity and position controllers from the Copter, Sub and Plane vehicle code into the AC_PosControl class.  This helps make the PosControl library more self contained.

- create a new AC_PID_2D class which is a 2-axis PID controller similar to the existing 2-axis PI controller
- replace the existing 2-axis PI velocity controller (instantiated in the Copter, Sub and Plane vehicle code) with a 2-axis PID controller and move into the PosControl library.
- move the remaining P and PID controllers from the vehicle class into the AC_PosControl library:
  - p_pos_xy (POS_XY_P) is the horizontal position P controller
  - p_alt_hold (POS_Z_P) is the vertical position P controller
  - p_vel_z (VEL_Z_P) is the vertical velocity P controller
  - pid_accel_z (ACCEL_Z_P, I, D, FILT) is the vertical acceleration PID controller

Existing parameter values are moved for Copter (see bottom of Parameters.cpp) but not for Sub or Plane.  I think for Sub this is not an issue because very few Sub users use position control.  @jaxxzer is this OK.

Plane does not seem to be able to move parameters within the quadplane object as far as I can tell.  Is this correct @tridge?

This change will require modifications to the Mission Planner (@meee1) extended tuning screen and the wiki will need to be updated (which I will take care of closer to the release)>

### CODE REVIEW

**Comment by rmackay9** [comment]
rebased on master but no additional changes made.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    // reset input filter to value received
+    if (_flags._reset_filter) {
+        _flags._reset_filter = false;
+        _input = input;
+    }
+
+    // update filter and calculate derivative
+    const Vector2f input_delta = (input - _input) * _filt_alpha;
+    _input = _input + input_delta;
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.cpp]: txs, fixed in follow up commit

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    set_input_filter_d(input_delta);
+}
+
+// set_input_filter_d - set input to PID controller
+//  only input to the D portion of the controller is filtered
+//  this should be called before any other calls to get_p, get_i or get_d
+void AC_PID_2D::set_input_filter_d(Vector2f input_delta)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.cpp]: yup, changed to const reference in follow-up commit, thanks!

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+{
+    // don't process inf or NaN
+    if (!isfinite(input_delta.x) && !isfinite(input_delta.y)) {
+        return;
+    }
+
+    // reset input filter to value received
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.cpp]: yup, fixed up comment

File: libraries/AC_PID/AC_PID_2D.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Param/AP_Param.h>
+#include <stdlib.h>
+#include <cmath>
+
+#define AC_PID_2D_FILT_HZ_DEFAULT  20.0f   // default input filter frequency
+#define AC_PID_2D_FILT_HZ_MIN      0.01f   // minimum input filter frequency
+#define AC_PID_2D_FILT_D_HZ_DEFAULT  10.0f   // default input filter frequency
+#define AC_PID_2D_FILT_D_HZ_MIN      0.005f   // minimum input filter frequency
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.h]: ok, moved definitions to .cpp in follow up commit.

File: libraries/AC_PID/AC_PID_2D.cpp [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    _kp.save();
+    _ki.save();
+    _kd.save();
+    _imax.save();
+    _filt_hz.save();
+    _filt_d_hz.save();
+}
+
+/// Overload the function call operator to permit easy initialisation
+void AC_PID_2D::operator() (float p, float i, float imaxval, float input_filt_hz, float input_filt_d_hz, float dt)
```
[the remark this answers is not in the document]
Comment (rmackay9) [reply libraries/AC_PID/AC_PID_2D.cpp]: ok, removed

**Comment by rmackay9** [comment]
@tridge has said he would like the parameter conversion for quadplane.  I need a bit of advice on how to do this because at the moment, I don't see how we can do parameter conversions for parameters that are not in the main parameter table.  I.e. parameters in g2 or the quadplane object.

**Comment by lthall** [comment]
Thanks @jaxxzer. When I get the rewrite of multi done we should have a chat about what I should be thinking about to make sure sub is supported properly.

**Comment by WickedShell** [comment]
@rmackay9 What happened with conversion code for quadplane?

**Comment by WickedShell** [comment]
I'm just worried that it's going to get pushed off and not handled, or left as a cleanup for someone else. In general we have refused to merge stuff that didn't include conversion code in the past/that didn't handle all vehicles correctly.

**Comment by rmackay9** [comment]
@WickedShell, I promise I will tackle it as my top priority once @tridge is back and can provide guidance.

**Pull request merged**